libsquoosh/src/missing-types.d.ts

```diff
@@ -0,0 +1,178 @@
+  }
+
+  var Global: {
+    prototype: Global;
+    new (descriptor: GlobalDescriptor, v?: any): Global;
+  };
+
+  interface Instance {
+    readonly exports: Exports;
+  }
+
+  var Instance: {
+    prototype: Instance;
+    new (module: Module, importObject?: Imports): Instance;
+  };
+
+  interface LinkError {}
+
+  var LinkError: {
+    prototype: LinkError;
+    new (): LinkError;
+  };
+
+  interface Memory {
+    readonly buffer: ArrayBuffer;
+    grow(delta: number): number;
+  }
+
+  var Memory: {
+    prototype: Memory;
+    new (descriptor: MemoryDescriptor): Memory;
+  };
+
+  interface Module {}
+
+  var Module: {
+    prototype: Module;
+    new (bytes: BufferSource): Module;
+    customSections(moduleObject: Module, sectionName: string): ArrayBuffer[];
+    exports(moduleObject: Module): ModuleExportDescriptor[];
+    imports(moduleObject: Module): ModuleImportDescriptor[];
+  };
+
+  interface RuntimeError {}
+
+  var RuntimeError: {
+    prototype: RuntimeError;
+    new (): RuntimeError;
+  };
+
+  interface Table {
+    readonly length: number;
+    get(index: number): Function | null;
+    grow(delta: number): number;
+    set(index: number, value: Function | null): void;
+  }
+
+  var Table: {
+    prototype: Table;
+    new (descriptor: TableDescriptor): Table;
+  };
+
+  interface GlobalDescriptor {
+    mutable?: boolean;
+    value: ValueType;
+  }
+
+  interface MemoryDescriptor {
+    initial: number;
+    maximum?: number;
+  }
+
+  interface ModuleExportDescriptor {
+    kind: ImportExportKind;
+    name: string;
+  }
+
+  interface ModuleImportDescriptor {
+    kind: ImportExportKind;
+    module: string;
+    name: string;
+  }
+
+  interface TableDescriptor {
+    element: TableKind;
+    initial: number;
+    maximum?: number;
+  }
+
+  interface WebAssemblyInstantiatedSource {
+    instance: Instance;
+    module: Module;
+  }
+
+  type ImportExportKind = 'function' | 'global' | 'memory' | 'table';
+  type TableKind = 'anyfunc';
+  type ValueType = 'f32' | 'f64' | 'i32' | 'i64';
+  type ExportValue = Function | Global | Memory | Table;
+  type Exports = Record<string, ExportValue>;
+  type ImportValue = ExportValue | number;
+  type ModuleImports = Record<string, ImportValue>;
+  type Imports = Record<string, ModuleImports>;
+  function compile(bytes: BufferSource): Promise<Module>;
+  // `compileStreaming` does not exist in NodeJS
+  // function compileStreaming(source: Response | Promise<Response>): Promise<Module>;
+  function instantiate(
+    bytes: BufferSource,
+    importObject?: Imports,
+  ): Promise<WebAssemblyInstantiatedSource>;
+  function instantiate(
+    moduleObject: Module,
+    importObject?: Imports,
+  ): Promise<Instance>;
+  // `instantiateStreaming` does not exist in NodeJS
+  // function instantiateStreaming(response: Response | PromiseLike<Response>, importObject?: Imports): Promise<WebAssemblyInstantiatedSource>;
+  function validate(bytes: BufferSource): boolean;
+}
```

Comment:
This shouldn’t be necessary. The WebAssembly Typings are available from the standard type library. Is it erroring for you?

Reply:
When I remove the WebAssembly definitions from here and run tsc through rollup with `build` script I get these errors:
```
../emscripten-types.d.ts:40:16 - error TS2503: Cannot find namespace 'WebAssembly'.

40       imports: WebAssembly.Imports,
                  ~~~~~~~~~~~

../emscripten-types.d.ts:41:33 - error TS2503: Cannot find namespace 'WebAssembly'.

41       successCallback: (module: WebAssembly.Module) => void,
                                   ~~~~~~~~~~~

../emscripten-types.d.ts:42:8 - error TS2503: Cannot find namespace 'WebAssembly'.

42     ): WebAssembly.Exports;
          ~~~~~~~~~~~
```

As far as I see, WebAssembly definitions exist in `lib.dom.d.ts` and `lib.webworker.d.ts` however (as far as I understand, again 😅 ) libSquoosh is currently aimed to be run in NodeJS runtime so we shouldn't use definitions from `lib.dom.d.ts` and `lib.worker.d.ts` but only from `esnext` and `@types/node`. So, essentially, this definitions must exist in `@types/node` but they currently do not exist hence we need to include them ourselves.

For just rewriting ImageData, we don't need those definitions actually so I'll remove this for now; but while we were rewriting codecs.js to codecs.ts we'll need definitions for emscripten which also needs definitions for WebAssembly. Because of this, I'll reintroduce this in that PR and we can discuss the details there :) (It looks like I shouldn't have included this file in this PR in the first place... 😄 )